pkg/controller/awsfederatedaccountaccess/awsfederatedaccountaccess_controller.go

```diff
@@ -0,0 +1,368 @@
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	// If the state is applied don't do anything
+	if currentFAA.Status.State == awsv1alpha1.AWSFederatedAccountStateReady || currentFAA.Status.State == awsv1alpha1.AWSFederatedAccountStateFailed {
+		return reconcile.Result{}, nil
+	}
+
+	// Get a list of all available roles
+	federatedRoleList := &awsv1alpha1.AWSFederatedRoleList{}
+	if r.client.List(context.TODO(), &client.ListOptions{}, federatedRoleList); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Create map to make finding role easier
+	roleMap := make(map[string]awsv1alpha1.AWSFederatedRole)
+	for _, role := range federatedRoleList.Items {
+		roleMap[role.Name] = role
+	}
+
+	// See if requested role exists
+	if _, ok := roleMap[currentFAA.Spec.AWSFederatedRoleName]; !ok {
+		SetStatuswithCondition(currentFAA, "Requested role does not exist", awsv1alpha1.AWSFederatedAccountFailed, awsv1alpha1.AWSFederatedAccountStateFailed)
+		reqLogger.Error(ErrFederatedAccessRoleNotFound, fmt.Sprintf("Resquested role '%s' not found", currentFAA.Spec.AWSFederatedRoleName))
+
+		err := r.client.Status().Update(context.TODO(), currentFAA)
+		if err != nil {
+			reqLogger.Error(err, fmt.Sprintf("Status update for %s failed", currentFAA.Name))
+			return reconcile.Result{}, err
+		}
+
+		return reconcile.Result{}, nil
+
+	}
+
+	// Get aws client
+	awsClient, err := awsclient.GetAWSClient(r.client, awsclient.NewAwsClientInput{
+		SecretName: currentFAA.Spec.AwsCredentialSecret.Name,
+		NameSpace:  currentFAA.Spec.AwsCredentialSecret.Namespace,
+		AwsRegion:  "us-east-1",
+	})
+	if err != nil {
+		awsClientErr := fmt.Sprintf("Unable to create aws client for region ")
+		reqLogger.Error(err, awsClientErr)
+		return reconcile.Result{}, err
+	}
+
+	// here create the custom policy in the cluster account
+	err = r.createIAMPolicy(awsClient, roleMap[currentFAA.Spec.AWSFederatedRoleName], *currentFAA)
+	if err != nil {
+		// if we were unable to create the policy fail this CR.
+		SetStatuswithCondition(currentFAA, "Failed to create custom policy", awsv1alpha1.AWSFederatedAccountFailed, awsv1alpha1.AWSFederatedAccountStateFailed)
+		reqLogger.Error(err, fmt.Sprintf("Unable to create policy resquested by '%s'", currentFAA.Name))
+
+		err := r.client.Status().Update(context.TODO(), currentFAA)
+		if err != nil {
+			reqLogger.Error(err, fmt.Sprintf("Status update for %s failed", currentFAA.Name))
+			return reconcile.Result{}, err
+		}
+
+		return reconcile.Result{}, nil
+
+	}
+
+	// create role role and apply custom policys and awsmanagedpolicies
+	err = r.createIAMRole(awsClient, roleMap[currentFAA.Spec.AWSFederatedRoleName], *currentFAA)
+	if err != nil {
+		SetStatuswithCondition(currentFAA, "Failed to create role", awsv1alpha1.AWSFederatedAccountFailed, awsv1alpha1.AWSFederatedAccountStateFailed)
```

Comment:
Again does API failure mean the CR should go into a failed state?

Reply:
Happy to leave these as is provided we track it after MVP.